contracts/ServiceNodeRewards.sol

```diff
@@ -1,167 +1,157 @@
     /// @param ids An array of service node IDs that did not sign and to be excluded from aggregation.
-	function updateRewardsBalance(
-		address recipientAddress, 
-		uint256 recipientAmount,
-		uint256 sigs0,
-		uint256 sigs1,
-		uint256 sigs2,
-		uint256 sigs3,
-		uint64[] memory ids
-	) public {
+	function updateRewardsBalance( address recipientAddress, uint256 recipientRewards, BLSSignatureParams calldata blsSignature, uint64[] memory ids) external whenNotPaused {
         if (!IsActive) revert ContractNotActive();
-        if (recipientAmount == 0 || recipientAddress == address(0)) revert InvalidParameter();
+        if (recipientAddress == address(0)) revert NullRecipient();
         if (ids.length > blsNonSignerThreshold) revert InsufficientBLSSignatures(serviceNodesLength() - ids.length, serviceNodesLength() - blsNonSignerThreshold);
+        if (recipients[recipientAddress].rewards >= recipientRewards) revert RecipientRewardsTooLow();
 		BN256G1.G1Point memory pubkey;
 		for(uint256 i = 0; i < ids.length; i++) {
-			pubkey = BN256G1.add(pubkey, serviceNodes[ids[i]].pubkey);
+			pubkey = BN256G1.add(pubkey, _serviceNodes[ids[i]].pubkey);
 		}
-		pubkey = BN256G1.add(aggregate_pubkey, BN256G1.negate(pubkey));
-		BN256G2.G2Point memory signature = BN256G2.G2Point([sigs1,sigs0],[sigs3,sigs2]);
-		bytes memory encodedMessage = abi.encodePacked(rewardTag, recipientAddress, recipientAmount);
+		pubkey = BN256G1.add(_aggregatePubkey, BN256G1.negate(pubkey));
+        BN256G2.G2Point memory signature = BN256G2.G2Point([blsSignature.sigs1,blsSignature.sigs0],[blsSignature.sigs3,blsSignature.sigs2]);
+		bytes memory encodedMessage = abi.encodePacked(rewardTag, recipientAddress, recipientRewards);
 		BN256G2.G2Point memory Hm = BN256G2.hashToG2(BN256G2.hashToField(string(encodedMessage)));
 		if (!Pairing.pairing2(BN256G1.P1(), signature, BN256G1.negate(pubkey), Hm)) revert InvalidBLSSignature();
 
 		uint256 previousBalance = recipients[recipientAddress].rewards;
-		recipients[recipientAddress].rewards = recipientAmount;
-		emit RewardsBalanceUpdated(recipientAddress, recipientAmount, previousBalance);
+		recipients[recipientAddress].rewards = recipientRewards;
+		emit RewardsBalanceUpdated(recipientAddress, recipientRewards, previousBalance);
 	}
 
-
-    /// @notice Builds a message for recipient reward calculation.
-    /// @param recipientAddress The address of the recipient.
-    /// @param balance The balance to be encoded in the message.
-    /// @return The encoded message.
-    function buildRecipientMessage(address recipientAddress, uint256 balance) public pure returns (bytes memory) {
-        return abi.encode(recipientAddress, balance);
-    }
-
     /// @dev Internal function to handle reward claims. Will transfer the available rewards worth of our token to claimingAddress
     /// @param claimingAddress The address claiming the rewards.
     function _claimRewards(address claimingAddress) internal {
```

Comment:
`claimRewards` which thunks into `_claimRewards` looks like it can be removed. `_claimRewards` is marked internal but no one else calls the function internally, the only function that calls it is `claimRewards` which can only pass in `msg.sender`

Instead we could just do

```solidity
function claimRewards() internal {
    uint256 claimedRewards = recipients[msg.sender].claimed;
    // e.t.c.
}
```

No need to have a wrapper function and would reduce LOC and consequently complexity.